File: src/ratelimit.ts
/**
 * @module ratelimit
 * Rate Limiter & Spending Cap Guardrails.
 *
 * Prevents runaway AI agents from executing an unlimited number of transactions
 * or spending more than a configured maximum within a rolling time window.
 *
 * These are pure in-memory guardrails (no RPC calls needed). They are designed
 * to be instantiated once and shared across the agent's lifecycle.
 */

export type RateLimitResult = {
  decision: "ALLOW" | "BLOCK";
  reason: string;
  allowExecute: boolean;
};

export interface RateLimiterConfig {
  /** Maximum number of transactions allowed within the window */
  maxTransactions: number;
  /** Time window in seconds (e.g. 3600 for 1 hour) */
  windowSeconds: number;
}

export interface SpendingCapConfig {
  /** Maximum cumulative value allowed within the window (in wei or smallest token unit) */
  maxSpend: bigint;
  /** Time window in seconds */
  windowSeconds: number;
}

/**
 * In-memory transaction rate limiter.
 * Tracks timestamps of recent transactions within a sliding window.
 */
export class RateLimiter {
  private readonly maxTx: number;
  private readonly windowMs: number;
  private timestamps: number[] = [];

  constructor(config: RateLimiterConfig) {
    if (typeof config.maxTransactions !== "number" || config.maxTransactions <= 0) {
      throw new Error("maxTransactions must be a positive integer");
    }
    if (typeof config.windowSeconds !== "number" || config.windowSeconds <= 0) {
      throw new Error("windowSeconds must be a positive number");
    }
    this.maxTx = config.maxTransactions;
    this.windowMs = config.windowSeconds * 1000;
  }

  /**
   * Call this BEFORE each transaction. Returns ALLOW if under the rate limit,
   * BLOCK if the agent has exceeded the transaction cap for this window.
   */
  check(): RateLimitResult {
    const now = Date.now();
    const cutoff = now - this.windowMs;

    // Prune expired timestamps
    this.timestamps = this.timestamps.filter((t) => t > cutoff);

    if (this.timestamps.length >= this.maxTx) {
      return {
        decision: "BLOCK",
        reason: `RATE LIMIT EXCEEDED: ${this.timestamps.length}/${this.maxTx} transactions in the last ${this.windowMs / 1000}s window. — BLOCK`,
        allowExecute: false,
      };
    }

    return {
      decision: "ALLOW",
      reason: `rate limit ok (${this.timestamps.length}/${this.maxTx})`,
      allowExecute: true,
    };
  }

  /**
   * Call this AFTER a transaction is successfully submitted to record it.
   */
  record(): void {
    this.timestamps.push(Date.now());
  }

  /** Returns how many transactions remain in the current window */
  remaining(): number {
    const cutoff = Date.now() - this.windowMs;
    this.timestamps = this.timestamps.filter((t) => t > cutoff);
    return Math.max(0, this.maxTx - this.timestamps.length);
  }
}

/**
 * In-memory spending cap enforcer.
 * Tracks cumulative spend within a sliding window.
 */
export class SpendingCap {
  private readonly maxSpend: bigint;
  private readonly windowMs: number;
  private ledger: Array<{ timestamp: number; amount: bigint }> = [];

  constructor(config: SpendingCapConfig) {
    if (typeof config.maxSpend !== "bigint" || config.maxSpend <= 0n) {
      throw new Error("maxSpend must be a positive bigint");
    }
    if (typeof config.windowSeconds !== "number" || config.windowSeconds <= 0) {
      throw new Error("windowSeconds must be a positive number");
    }
    this.maxSpend = config.maxSpend;
    this.windowMs = config.windowSeconds * 1000;
  }

  /**
   * Check if the proposed spend would exceed the cap.
   * @param proposedAmount The value of the transaction the agent wants to execute (in wei)
   */
  check(proposedAmount: bigint): RateLimitResult {
    if (typeof proposedAmount !== "bigint" || proposedAmount < 0n) {
      return {
        decision: "BLOCK",
        reason: "invalid proposedAmount — BLOCK",
        allowExecute: false,
      };
    }

    const now = Date.now();
    const cutoff = now - this.windowMs;
    this.ledger = this.ledger.filter((e) => e.timestamp > cutoff);

    const currentSpend = this.ledger.reduce((sum, e) => sum + e.amount, 0n);
    const projectedSpend = currentSpend + proposedAmount;

    if (projectedSpend > this.maxSpend) {
      return {
        decision: "BLOCK",
        reason: `SPENDING CAP EXCEEDED: projected spend ${projectedSpend} exceeds cap ${this.maxSpend} in the last ${this.windowMs / 1000}s window. — BLOCK`,
        allowExecute: false,
      };
    }

    return {
      decision: "ALLOW",
      reason: `spending cap ok (${currentSpend}/${this.maxSpend})`,
      allowExecute: true,
    };
  }

  /**
   * Record a successfully submitted transaction amount.
   */
  record(amount: bigint): void {
    this.ledger.push({ timestamp: Date.now(), amount });
  }

  /** Returns the remaining budget in the current window */
  remaining(): bigint {
    const cutoff = Date.now() - this.windowMs;
    this.ledger = this.ledger.filter((e) => e.timestamp > cutoff);
    const spent = this.ledger.reduce((sum, e) => sum + e.amount, 0n);
    const rem = this.maxSpend - spent;
    return rem > 0n ? rem : 0n;
  }
}

File: src/ratelimit.rs
//! In-memory rolling-window resource governors.
//!
//! [`RateLimiter`] caps transaction *frequency*; [`SpendingCap`] caps
//! cumulative *value*. Both are fail-closed advisory guards — see caveats.
//!
//! # Correctness caveats (read before relying on these)
//! - **Not thread-safe across tasks:** methods take `&mut self`. Share via
//!   `Mutex`/`RwLock` in concurrent agents.
//! - **`check` + `record` is check-then-act:** two concurrent `check()` calls
//!   can both pass before either `record()`s. Prefer the atomic
//!   [`RateLimiter::try_acquire`] / [`SpendingCap::try_spend`].
//! - **Volatile:** state lives in RAM and resets on restart. A restart
//!   clears the window — persist externally if restart-bypass matters.
//! - **Bounded:** histories are capped at [`MAX_HISTORY`] entries; excess
//!   oldest entries are evicted first. Eviction errs fail-OPEN (fewer
//!   recorded tx/spend reads as more headroom) — the cap is a DoS bound,
//!   not a safety bound; window semantics are approximate under flood.

use crate::types::GuardrailResult;
use std::time::{Duration, Instant};

/// Hard bound on retained timestamps/ledger entries (DoS cap).
pub const MAX_HISTORY: usize = 100_000;

/// Rolling-window transaction frequency governor.
///
/// ```rust
/// use stale::ratelimit::RateLimiter;
/// let mut limiter = RateLimiter::new(10, 60).unwrap();
/// assert!(limiter.try_acquire().allow_execute);
/// ```
pub struct RateLimiter {
    max_tx: usize,
    window: Duration,
    timestamps: Vec<Instant>,
    /// Set when flood eviction drops history. While set, `check` BLOCKs:
    /// evicted (possibly unexpired) entries would otherwise read as
    /// headroom (fail-open). Cleared once the window fully drains.
    evicted: bool,
}

impl RateLimiter {
    /// Create a limiter allowing `max_tx` transactions per `window_seconds`.
    ///
    /// # Errors
    /// `Err` if `max_tx == 0` or `window_seconds == 0`. `Err` if `max_tx`
    /// exceeds [`MAX_HISTORY`]: the retained history could never reach the
    /// policy, so the limiter would ALLOW forever under flood.
    pub fn new(max_tx: usize, window_seconds: u64) -> Result<Self, String> {
        if max_tx == 0 {
            return Err("max_tx must be greater than 0".to_string());
        }
        if max_tx > MAX_HISTORY {
            return Err(format!(
                "max_tx {} exceeds history bound {} — policy unenforceable — BLOCK (fail closed)",
                max_tx, MAX_HISTORY
            ));
        }
        if window_seconds == 0 {
            return Err("window_seconds must be greater than 0".to_string());
        }

        Ok(Self {
            max_tx,
            window: Duration::from_secs(window_seconds),
            timestamps: Vec::new(),
            evicted: false,
        })
    }

    /// Admission test. Takes `&mut self` to prune expired entries
    /// (housekeeping, not a verdict change). Prefer [`try_acquire`](Self::try_acquire)
    /// in production: `check` followed by a separate `record` is TOCTOU —
    /// a caller that forgets `record()` is never limited.
    pub fn check(&mut self) -> GuardrailResult {
        let now = Instant::now();
        let window = self.window;
        self.timestamps.retain(|&t| now.duration_since(t) < window);
        if self.timestamps.is_empty() {
            self.evicted = false;
        } else if self.evicted {
            return GuardrailResult::block(
                "rate history evicted under flood; headroom unverifiable — BLOCK (fail closed)",
            );
        }

        if self.timestamps.len() >= self.max_tx {
            GuardrailResult::block(format!(
                "RATE LIMIT EXCEEDED: {}/{} transactions in the last {}s window. — BLOCK",
                self.timestamps.len(),
                self.max_tx,
                self.window.as_secs()
            ))
        } else {
            GuardrailResult::allow(format!(
                "rate limit ok ({}/{})",
                self.timestamps.len(),
                self.max_tx
            ))
        }
    }

    /// Record a transaction (manual mode). Callers using [`check`](Self::check)
    /// must call this on every allowed transaction; otherwise the limiter
    /// never fills. Prefer [`try_acquire`](Self::try_acquire).
    pub fn record(&mut self) {
        self.timestamps.push(Instant::now());
        if self.timestamps.len() > MAX_HISTORY {
            let excess = self.timestamps.len() - MAX_HISTORY;
            self.timestamps.drain(..excess);
            self.evicted = true;
        }
    }

    /// Atomic check-and-record. Returns `Allow` (slot consumed) or `Block`
    /// (no slot consumed). This is the recommended entry point.
    pub fn try_acquire(&mut self) -> GuardrailResult {
        let verdict = self.check();
        if verdict.allow_execute {
            self.record();
        }
        verdict
    }

    /// Slots remaining in the current window. Takes `&mut self` to prune
    /// expired entries before counting.
    pub fn remaining(&mut self) -> usize {
        let now = Instant::now();
        let window = self.window;
        self.timestamps.retain(|&t| now.duration_since(t) < window);
        self.max_tx.saturating_sub(self.timestamps.len())
    }
}

/// Rolling-window cumulative spending governor (saturating `u128` math).
pub struct SpendingCap {
    max_spend: u128,
    window: Duration,
    ledger: Vec<(Instant, u128)>,
    /// Same flood-eviction semantics as [`RateLimiter::evicted`].
    evicted: bool,
}

impl SpendingCap {
    /// Create a cap allowing `max_spend` base units per `window_seconds`.
    ///
    /// # Errors
    /// `Err` if `max_spend == 0` or `window_seconds == 0`.
    pub fn new(max_spend: u128, window_seconds: u64) -> Result<Self, String> {
        if max_spend == 0 {
            return Err("max_spend must be greater than 0".to_string());
        }
        if window_seconds == 0 {
            return Err("window_seconds must be greater than 0".to_string());
        }

        Ok(Self {
            max_spend,
            window: Duration::from_secs(window_seconds),
            ledger: Vec::new(),
            evicted: false,
        })
    }

    /// Admission test for `proposed_amount`. Takes `&mut self` to prune
    /// expired entries (housekeeping). Same TOCTOU caveat as
    /// [`RateLimiter::check`]: prefer [`try_spend`](Self::try_spend).
    pub fn check(&mut self, proposed_amount: u128) -> GuardrailResult {
        let now = Instant::now();
        let window = self.window;
        self.ledger.retain(|&(t, _)| now.duration_since(t) < window);
        if self.ledger.is_empty() {
            self.evicted = false;
        } else if self.evicted {
            return GuardrailResult::block(
                "spend history evicted under flood; spend unverifiable — BLOCK (fail closed)",
            );
        }

        let current_spend: u128 = self
            .ledger
            .iter()
            .fold(0u128, |acc, &(_, amt)| acc.saturating_add(amt));

        let projected_spend = match current_spend.checked_add(proposed_amount) {
            Some(sum) => sum,
            None => return GuardrailResult::block("overflow in projected spend — BLOCK"),
        };

        if projected_spend > self.max_spend {
            GuardrailResult::block(format!(
                "SPENDING CAP EXCEEDED: projected spend {} exceeds cap {} in the last {}s window. — BLOCK",
                projected_spend, self.max_spend, self.window.as_secs()
            ))
        } else {
            GuardrailResult::allow(format!(
                "spending cap ok ({}/{})",
                current_spend, self.max_spend
            ))
        }
    }

    /// Record spend (manual mode). See [`try_spend`](Self::try_spend).
    pub fn record(&mut self, amount: u128) {
        self.ledger.push((Instant::now(), amount));
        if self.ledger.len() > MAX_HISTORY {
            let excess = self.ledger.len() - MAX_HISTORY;
            self.ledger.drain(..excess);
            self.evicted = true;
        }
    }

    /// Atomic check-and-record. Returns `Allow` (amount booked) or `Block`
    /// (nothing booked). Recommended entry point.
    pub fn try_spend(&mut self, amount: u128) -> GuardrailResult {
        let verdict = self.check(amount);
        if verdict.allow_execute {
            self.record(amount);
        }
        verdict
    }

    /// Spendable remainder in the current window (saturating). Takes
    /// `&mut self` to prune expired entries before summing.
    pub fn remaining(&mut self) -> u128 {
        let now = Instant::now();
        let window = self.window;
        self.ledger.retain(|&(t, _)| now.duration_since(t) < window);
        let spent: u128 = self
            .ledger
            .iter()
            .fold(0u128, |acc, &(_, amt)| acc.saturating_add(amt));
        self.max_spend.saturating_sub(spent)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_rate_limiter() {
        let mut limiter = RateLimiter::new(2, 60).unwrap();
        assert!(limiter.check().allow_execute);
        limiter.record();
        assert!(limiter.check().allow_execute);
        limiter.record();
        assert!(!limiter.check().allow_execute);
        assert_eq!(limiter.remaining(), 0);
    }

    #[test]
    fn test_spending_cap() {
        let mut cap = SpendingCap::new(1000, 60).unwrap();
        assert!(cap.check(400).allow_execute);
        cap.record(400);
        assert_eq!(cap.remaining(), 600);
        assert!(!cap.check(700).allow_execute);
        assert!(cap.check(600).allow_execute);
    }

    #[test]
    fn test_try_acquire_atomic() {
        let mut limiter = RateLimiter::new(1, 60).unwrap();
        assert!(limiter.try_acquire().allow_execute);
        // Slot consumed atomically — second acquire BLOCKs without manual record.
        assert!(!limiter.try_acquire().allow_execute);
    }

    #[test]
    fn test_try_spend_atomic() {
        let mut cap = SpendingCap::new(1000, 60).unwrap();
        assert!(cap.try_spend(600).allow_execute);
        assert!(!cap.try_spend(500).allow_execute);
        assert!(cap.try_spend(400).allow_execute);
    }

    #[test]
    fn test_oversized_policy_rejected_at_construction() {
        // A policy the history can never reach would ALLOW forever.
        assert!(RateLimiter::new(MAX_HISTORY + 1, 60).is_err());
        assert!(RateLimiter::new(MAX_HISTORY, 60).is_ok());
    }

    #[test]
    fn test_flood_eviction_blocks_fail_closed() {
        let mut limiter = RateLimiter::new(10, 60).unwrap();
        for _ in 0..(MAX_HISTORY + 5) {
            limiter.record();
        }
        // History was evicted: headroom unverifiable → BLOCK.
        assert!(!limiter.check().allow_execute);

        let mut cap = SpendingCap::new(u128::MAX, 60).unwrap();
        for _ in 0..(MAX_HISTORY + 5) {
            cap.record(1);
        }
        assert!(!cap.check(0).allow_execute);
    }
}
